common/beerocks/bwl/common/key_value_parser.cpp

```diff
@@ -0,0 +1,225 @@
+{
+    // Eliminate event log level from the begining of the event string : "<3>".
+    size_t idx_start = event_buffer_start_process_idx(event_str);
+
+    // An event string might start with keyless values such as Even Name.
+    // Find where the keyless parameters ends by finding where is the first parameter with key,
+    // assuming it has a preceding " <SomeKeyString>='".
+    auto idx = find_first_of_delimiter_pair(event_str, idx_start, ' ', '=');
+
+    // Put null terminator at the end of our key=val, for string stream construction.
+    if (idx != std::string::npos) {
+        event_str[idx] = '\0';
+    }
+
+    // Parse keyless params part and insert to the 'parsed_line' map object.
+    parse_event_keyless_params(event_str, idx_start, parsed_line);
+
+    // Fill the map with the rest of event data.
+    while (idx != std::string::npos) {
+
+        idx_start = ++idx;
+
+        // Find first '=' to skip on it
+        idx = event_str.find_first_of('=', idx_start);
+
+        // Find the next pair of delimiters index.
+        idx = find_first_of_delimiter_pair(event_str, ++idx, ' ', '=');
+
+        if (idx != std::string::npos) {
+            // Put null terminator at the end of our key=val, for ss_in construction.
+            event_str[idx] = '\0';
+        }
+
+        // Parse key=value
+        std::stringstream ss_in(event_str.c_str() + idx_start);
+        parse_line(ss_in, {'='}, parsed_line);
+    }
+}
+
+size_t KeyValueParser::event_buffer_start_process_idx(const std::string &event_str)
+{
+    // Eliminate event log level from the begining of the event string : "<3>"
+    auto idx_start = event_str.find_first_of(">");
+    if (idx_start != std::string::npos) {
+        idx_start++;
+    } else {
+        LOG(WARNING) << "Empty event! event_string: " << event_str;
+    }
+    return idx_start;
+}
+
+void KeyValueParser::parse_event_keyless_params(const std::string &event_str, size_t idx_start,
+                                                parsed_line_t &parsed_line)
+{
+    // Insert to map known prams without key
+    std::stringstream ss(event_str.c_str() + idx_start);
+    std::string str_storage;
+    bool opcode = true;
+    while (std::getline(ss, str_storage, ' ')) {
+        if (opcode) {
+            // assume that the second param is data or event name
+            parsed_line[EVENT_KEYLESS_PARAM_OPCODE] = str_storage;
+            opcode                                  = false;
+        } else if (beerocks::net::network_utils::is_valid_mac(str_storage)) {
+            parsed_line[EVENT_KEYLESS_PARAM_MAC] = str_storage;
+        } else if (!strncmp(str_storage.c_str(), "wlan", 4)) {
+            parsed_line[EVENT_KEYLESS_PARAM_IFACE] = str_storage;
+        }
+    }
+}
+
+bool KeyValueParser::read_param_int(const std::string &key, parsed_line_t &obj, int64_t &value,
+                                    bool ignore_unknown)
+{
```

Comment:
This would be better as
```
bool parsed_line_t::get_param(const std::string &key, int64_t &value, bool ignore_unknown);
```
(when we have this object-oriented stuff, we could just as well use it :-P

Reply:
What? `parsed_line_t` is an `unordered_map`. C++ doesn't support extension functions.